examples/tls-rustls/src/main.rs

```diff
@@ -42,3 +64,38 @@ async fn main() {
 async fn handler() -> &'static str {
     "Hello, World!"
 }
+
+async fn redirect_http_to_https(ports: Ports) {
+    fn make_https(host: String, uri: Uri, ports: Ports) -> Result<Uri, BoxError> {
+        let mut parts = uri.into_parts();
+
+        parts.scheme = Some(axum::http::uri::Scheme::HTTPS);
+
+        if parts.path_and_query.is_none() {
+            parts.path_and_query = Some("/".parse().unwrap());
+        }
+
+        let https_host = host.replace(&ports.http.to_string(), &ports.https.to_string());
+        parts.authority = Some(https_host.parse()?);
+
+        Ok(Uri::from_parts(parts)?)
+    }
+
+    let redirect = move |Host(host): Host, uri: Uri| async move {
+        match make_https(host, uri, ports) {
+            Ok(uri) => Ok(Redirect::permanent(&uri.to_string())),
+            Err(error) => {
+                tracing::warn!(%error, "failed to convert URI to HTTPS");
+                Err(StatusCode::BAD_REQUEST)
+            }
+        }
+    };
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], ports.http));
+    tracing::debug!("http redirect listening on {}", addr);
+
+    axum::Server::bind(&addr)
+        .serve(redirect.into_make_service())
+        .await
+        .unwrap()
+}
```

Comment:
I suggest depending on `url` crate and:

```rust
async fn redirect_http_to_https(ports: Ports) {
    let addr = SocketAddr::from(([127, 0, 0, 1], ports.http));
    tracing::debug!("http redirect listening on {}", addr);

    let app = redirect.layer(Extension(ports));

    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap()
}

async fn redirect(
    Host(host): Host,
    uri: Uri,
    Extension(ports): Extension<Ports>,
) -> Result<Redirect, StatusCode> {
    let path_and_query = uri
        .path_and_query()
        .map(ToString::to_string)
        .unwrap_or("".to_owned());

    let mut url = Url::parse(&format!("http://{}{}", host, path_and_query))
        .map_err(|_| StatusCode::BAD_REQUEST)?;

    url.set_scheme("https").unwrap();
    let _ = url.set_port(Some(ports.https));

    Ok(Redirect::permanent(&url.to_string()))
}
```

Reply:
What advantage does that give? The code doesn't seem less complicated than the version using `http::Uri` 🤔

Reply:
The version using `url` crate has a more functional approach, which may be preferred by some developers.  As a teaching example, though, I like that the version using `http::Uri` show how to work with the parts of `Uri` explicitly. I also prefer not to have to rely on an additional crate and middleware layer and extension extractor.

Reply:
I think this is simpler and less error prone. Things are more likely to go wrong when using `http::Uri`.

Reply:
I don't think pulling in an extra dependency for an example is worth it.